presto-main/src/main/java/com/facebook/presto/operator/ExchangeOperator.java

```diff
@@ -182,8 +187,13 @@ public Page getOutput()
         if (page == null) {
             return null;
         }
-
-        operatorContext.recordRawInput(page.getSizeInBytes(), page.getPositionCount());
+        int rawInputBytes = page.getSizeInBytes();
```

Comment:
I'm seeing that there are two exchange operators: ExchnageOperator and MergeOperator.

I'm also seeing that stats are aggregated to the query-level in com.facebook.presto.execution.QueryStats#create, where runtime stats are aggregates only within a stage, but not across stages.

```
// We prepend each metric name with the stage id to avoid merging metrics across stages.
```

Since we want aggregates shuffle stats at the query level, we can add extra logic to QueryStats#create.

There is logic to aggregate leaf stage input sizes:

```
                if (!plan.getTableScanSchedulingOrder().isEmpty()) {
                    rawInputDataSize += stageExecutionStats.getRawInputDataSize().toBytes();
                    rawInputPositions += stageExecutionStats.getRawInputPositions();
```

We can also aggregate rawInputXxx across non-leaf stages, which will give us the shuffle size / count.

```
                if (!plan.getTableScanSchedulingOrder().isEmpty()) {
                    rawInputDataSize += stageExecutionStats.getRawInputDataSize().toBytes();
                    rawInputPositions += stageExecutionStats.getRawInputPositions();
                  ...
                } else {
                    shuffleInputDataSize += stageExecutionStats.getRawInputDataSize().toBytes();
                    shuffleInputPositions += stageExecutionStats.getRawInputPositions();
                }
```

Reply:
Good point!